cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,255 @@
+	}
+	for _, tag := range toIgnore.List() {
+		tagsToCheck.Delete(tag)
+	}
+	if len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	return tagsToCheck, nil
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := newOpts()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	abs, err := filepath.Abs(opts.ciOperatorconfigPath)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to determine absolute CI Operator configuration path")
+	}
+	var promotedTags []api.ImageStreamTagReference
+	if err := config.OperateOnCIOperatorConfigDir(abs, func(cfg *api.ReleaseBuildConfiguration, metadata *config.Info) error {
+		promotedTags = append(promotedTags, release.PromotedTags(cfg)...)
+		return nil
+	}); err != nil {
+		logrus.WithField("path", abs).Fatal("failed to operate on CI Operator's config directory")
+	}
+
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", opts.kubeconfig)
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not load kube config from path %s", opts.kubeconfig)
+	}
+
+	client, err := ctrlruntimeclient.New(kubeConfig, ctrlruntimeclient.Options{})
+	if err != nil {
+		logrus.WithError(err).Fatalf("could not create client")
+	}
+
+	if err := validatePromotedTags(promotedTags); err != nil {
+		logrus.WithError(err).Fatalf("could not validate promoted tags")
+	}
+
+	toDelete, err := tagsToDelete(context.TODO(), client, promotedTags, opts.ignoredImageStreamTags)
+	if err != nil {
+		logrus.WithError(err).Fatal("could not get tags to delete")
+	}
+
+	if opts.openshiftMappingConfigPath != "" {
+		mappings := generateMappings(promotedTags, opts.openshiftMappingConfig)
+		for filename, mapping := range mappings {
+			var b bytes.Buffer
+			for src, targets := range mapping {
+				b.WriteString(fmt.Sprintf("%s\n", strings.Join(append([]string{src}, targets...), " ")))
+			}
+			f := filepath.Join(opts.openshiftMappingDir, filename)
+			logrus.WithField("filename", f).Info("Writing to file")
+			if err := ioutil.WriteFile(f, b.Bytes(), 0644); err != nil {
+				logrus.WithError(err).WithField("filename", f).Fatal("could not write to file")
+			}
+		}
+	}
+
+	if opts.dryRun {
+		logrus.WithField("Len", toDelete.Len()).WithField("List", toDelete.List()).Info("Tags to delete")
+	} else {
+		logrus.Info("TODO: delete tags for real")
+	}
+}
+
+func generateMappings(promotedTags []api.ImageStreamTagReference, mappingConfig *OpenshiftMappingConfig) map[string]map[string][]string {
+	mappings := map[string]map[string][]string{}
+	for version := range mappingConfig.Images {
+		mappings[fmt.Sprintf("mapping_origin_%s", strings.ReplaceAll(version, ".", "_"))] = map[string][]string{}
+	}
+	for _, tag := range promotedTags {
+		if tag.Namespace == mappingConfig.SourceNamespace {
```

Comment:
This is what causes your tool to delete all the mapping configs. Let's focus on the ImageStreamTags for now.

Reply:
Removed the logic of the mappings.